### PR TITLE
Resize View without Initialization

### PR DESCRIPTION
**Pull request by vqd8a**
Address issue #2048 .

### CODE REVIEW

**Comment by dalg24** [comment]
Add to whitelist

File: core/unit_test/TestResize.hpp [hunk 1]
```diff
@@ -132,8 +132,339 @@ void testResize() {
...
+    view_type view_8d("view_8d", sizes[0], sizes[1], sizes[2], sizes[3],
+                      sizes[4], sizes[5], sizes[6], sizes[7]);
+    const int* oldPointer = view_8d.data();
+    EXPECT_TRUE(oldPointer != NULL);
+    Kokkos::resize(Kokkos::view_resize(Kokkos::WithoutInitializing), view_8d,
+                   sizes[0], sizes[1], sizes[2], sizes[3], sizes[4], sizes[5],
+                   sizes[6], sizes[7]);
+    const int* newPointer = view_8d.data();
+    EXPECT_TRUE(oldPointer == newPointer);
+  }
```
Comment (dalg24) [inline core/unit_test/TestResize.hpp]: This block (above) is pretty much duplicated (only change is injection of `Kokkos::view_resize(Kokkos::WithoutInitializing)` as 1st parameter passed to `Kokkos::resize()`

File: core/unit_test/TestResize.hpp [hunk 2]
```diff
@@ -132,8 +132,339 @@ void testResize() {
...
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    EXPECT_TRUE(test == true);
+  }
```
Comment (dalg24) [inline core/unit_test/TestResize.hpp]: This block isn't even though it is something we should have been testing for as well in the default case with data initialization.

File: core/unit_test/TestResize.hpp (hunk 1 above)
Comment (vqd8a) [reply core/unit_test/TestResize.hpp]: @dalg24. Thanks for your comments. In the new commit, I added tests in to CMakeLists.txt and combine the ```WithoutInitializing``` tests with the default ```resize``` tests.

File: core/unit_test/TestResize.hpp (hunk 2 above)
Comment (vqd8a) [reply core/unit_test/TestResize.hpp]: @dalg24, thanks. The default case is tested in my new commit.

File: core/perf_test/PerfTest_ViewResize.hpp [hunk 3]
```diff
@@ -222,6 +296,16 @@ void run_resizeview_tests7(int N, int R) {
     }
     time7 = timer.seconds() / R;
   }
+  {
+    Kokkos::View<double*******, Layout> a("A7", N2, N1, N1, N1, N1, N1, N1);
+    Kokkos::Timer timer;
+    for (int r = 0; r < R; r++) {
+      Kokkos::View<double*******, Layout> a_(a);
+      Kokkos::resize(Kokkos::view_resize(Kokkos::WithoutInitializing), a_,
+                     int(N2 * 1.1), N1, N1, N1, N1, N1, N1);
```
Comment (Rombur) [inline core/perf_test/PerfTest_ViewResize.hpp]: Is there a reason why you always only resize the first dimension? Also did you make sure that `N1`, `N2`, etc. are large enough that `int(N2 * 1.1)` is greater than `N2`?

File: core/unit_test/TestResize.hpp [hunk 4]
```diff
@@ -40,25 +40,29 @@
 // ************************************************************************
 //@HEADER
 */
-#ifndef TESTVIEWSUBVIEW_HPP_
-#define TESTVIEWSUBVIEW_HPP_
+#ifndef TESTRESIZE_HPP_
+#define TESTRESIZE_HPP_
```
Comment (Rombur) [inline core/unit_test/TestResize.hpp]: Nice catch

File: core/unit_test/CMakeLists.txt [hunk 5]
```diff
@@ -414,6 +417,7 @@ IF(Kokkos_ENABLE_CUDA)
       cuda/TestCuda_Crs.cpp
       cuda/TestCuda_WorkGraph.cpp
       cuda/TestCuda_UniqueToken.cpp
+      cuda/TestCuda_ViewResize.cpp
```
Comment (Rombur) [inline core/unit_test/CMakeLists.txt]: We should also test this for the hpx backend

File: core/perf_test/PerfTest_ViewResize.hpp (hunk 3 above)
Comment (vqd8a) [reply core/perf_test/PerfTest_ViewResize.hpp]: @Rombur Thanks, Bruno.
I just inherited the perf test for the default resize (i.e. with initialization ) here. But I think with only the first dimension we can test both LayoutLeft (stride of 1) and LayoutRight (stride > 1) for multiple dimensions. @crtrott can you correct me if I am wrong?

I already checked int(N1*1.1) is greater than N1, int(N2*1.1) is greater than N2, etc.

File: core/unit_test/TestResize.hpp (hunk 4 above)
Comment (vqd8a) [reply core/unit_test/TestResize.hpp]: @Rombur Thanks, Bruno.

File: core/unit_test/CMakeLists.txt (hunk 5 above)
Comment (vqd8a) [reply core/unit_test/CMakeLists.txt]: @Rombur I added the test for the hpx backend. Thanks.

File: core/src/Kokkos_CopyViews.hpp [hunk 6]
```diff
@@ -2895,6 +2895,103 @@ resize(Kokkos::View<T, P...>& v, const size_t n0 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
...
+  // an optimization.
+
+  // TODO (mfh 27 Jun 2017) If the old View has enough space but just
+  // different dimensions (e.g., if the product of the dimensions,
+  // including extra space for alignment, will not change), then
+  // consider just reusing storage.  For now, Kokkos always
+  // reallocates if any of the dimensions change, even if the old View
+  // has enough space.
+
+  view_type v_resized(view_alloc (v.label(), WithoutInitializing), n0, n1, n2, n3, n4, n5, n6, n7);
```
[the remark this answers is not in the document]
Comment (vqd8a) [reply core/src/Kokkos_CopyViews.hpp]: @crtrott I fixed it and simplified the interface a bit (not using ```view_resize(...)```). Not sure if this is what you wanted.

**Comment by vqd8a** [comment]
Thanks, @dalg24  Damien for editing the unit tests. I pushed some edits to address Bruno and Christian's comments.

File: core/perf_test/PerfTest_ViewResize.hpp (hunk 3 above)
Comment (dalg24) [reply core/perf_test/PerfTest_ViewResize.hpp]: @ctrott please confirm and merge if you are happy with the PR.

**Pull request merged**